kv/kvnemesis: TestKVNemesisMultiNode failed

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11202372?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11202372?buildTab=artifacts#/) on master @ [695d4ec85157a4da0c97237e6f16bf919eebf6c5](https://github.com/cockroachdb/cockroach/commits/695d4ec85157a4da0c97237e6f16bf919eebf6c5):

Fatal error:

```
panic: use of Span after Finish. Span: /cockroach.roachpb.Internal/Batch. Finish previously called at: <stack not captured. Set debugUseAfterFinish> [recovered]
	panic: use of Span after Finish. Span: /cockroach.roachpb.Internal/Batch. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

```
Stack: 

```
goroutine 6569871 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x1f81c6a?, {0x6976bc8, 0xc014ecf170})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x6a
panic({0x4b2d840, 0xc01b174160})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0x6976bc8?)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:182 +0x12e
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Tracer(...)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:225
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaDecoder).createTracingSpans(0xc00a3270d8, {0x6976bc8, 0xc0204de0f0})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_application_decoder.go:154 +0x4f0
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaDecoder).DecodeAndBind(0xc00a326c80?, {0x6976bc8, 0xc0204de0f0}, {0xc02851e210?, 0x3?, 0x10?})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_application_decoder.go:63 +0x66
github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.(*Task).Decode(...)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply/task.go:142
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:878 +0x73a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:728 +0x1b0
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady(0xc0165eea80, 0x2?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:689 +0x14b
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker(0xc0085cbd50, {0x6976bc8, 0xc014ecf170}, {0x695c980, 0xc0165eea80}, 0xc0165ee000)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:416 +0x274
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2({0x6976bc8?, 0xc014ecf170?})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:321 +0x45
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x146
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x43b
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestKVNemesisMultiNode
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode1784560352
    test_log_scope.go:81: use -show-logs to present logs inline
    kvnemesis_test.go:283: seed: 2381536854787156731
    test_server_shim.go:98: 
        Test server was configured to route SQL queries to a secondary tenant (virtual cluster).
        If you are only seeing a test failure when this message appears, there may be a problem
        specific to cluster virtualization or multi-tenancy.
        
        To investigate, consider using "COCKROACH_TEST_TENANT=true" to force-enable just
        the secondary tenant in all runs (or, alternatively, "false" to force-disable), or use
        "COCKROACH_INTERNAL_DISABLE_METAMORPHIC_TESTING=false" to disable all random test variables altogether.
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
    kvnemesis_test.go:197: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1734753383
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30377

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11311374?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11311374?buildTab=artifacts#/) on master @ [c13bf7633cbb416d9e43f8c57b1e309fab1110ce](https://github.com/cockroachdb/cockroach/commits/c13bf7633cbb416d9e43f8c57b1e309fab1110ce):

Fatal error:

```
panic: use of Span after Finish. Span: /cockroach.roachpb.Internal/Batch. Finish previously called at: <stack not captured. Set debugUseAfterFinish> [recovered]
	panic: use of Span after Finish. Span: /cockroach.roachpb.Internal/Batch. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

```
Stack: 

```
goroutine 8388484 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x1f8976a?, {0x69b21a8, 0xc008aa3ce0})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x6a
panic({0x4b58900, 0xc01654c060})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0x9ac8c70?)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:182 +0x12e
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).RecordingType(0xc0028ca400)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:442 +0x1e
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).IsVerbose(...)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:450
github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog({0x69b21a8, 0xc009411bc0})
	github.com/cockroachdb/cockroach/pkg/util/log/trace.go:92 +0x3a
github.com/cockroachdb/cockroach/pkg/util/log.vEventf({0x69b21a8, 0xc009411bc0}, 0x0, 0x1, 0x0?, 0x0?, {0x502c335, 0x19}, {0x0, 0x0, ...})
	github.com/cockroachdb/cockroach/pkg/util/log/trace.go:212 +0x11f
github.com/cockroachdb/cockroach/pkg/util/log.VEvent(...)
	github.com/cockroachdb/cockroach/pkg/util/log/trace.go:231
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*propBuf).FlushLockedWithRaftGroup(0xc0103354a0, {0x69b21a8, 0xc02912cd20}, {0x69d0c18, 0xc019afa230})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_proposal_buf.go:601 +0x12c8
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked.func2(0x4893b00?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:775 +0xd4
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).withRaftGroupLocked(0xc010334c80, 0x4afa240?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1998 +0x4e
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:772 +0x317
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:728 +0x1b0
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady(0xc004b6ca80, 0x69b21a8?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:689 +0x14b
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker(0xc00bb7a850, {0x69b21a8, 0xc008aa3ce0}, {0x6997e70, 0xc004b6ca80}, 0xc004d98000)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:418 +0x195
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2({0x69b21a8?, 0xc008aa3ce0?})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:321 +0x45
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x146
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x43b
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestKVNemesisMultiNode
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode628115778
    test_log_scope.go:81: use -show-logs to present logs inline
    kvnemesis_test.go:283: seed: 4653790105750483978
    test_server_shim.go:115: 
        Test server was configured to route SQL queries to a secondary tenant (virtual cluster).
        If you are only seeing a test failure when this message appears, there may be a problem
        specific to cluster virtualization or multi-tenancy.
        
        To investigate, consider using "COCKROACH_TEST_TENANT=true" to force-enable just
        the secondary tenant in all runs (or, alternatively, "false" to force-disable), or use
        "COCKROACH_INTERNAL_DISABLE_METAMORPHIC_TESTING=false" to disable all random test variables altogether.
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
    kvnemesis_test.go:197: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2755337404
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/replica_application_result.go
+++ b/pkg/kv/kvserver/replica_application_result.go
@@ -266,22 +266,21 @@ func (r *Replica) prepareLocalResult(ctx context.Context, cmd *replicatedCmd) {
 	}
 }
 
-func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *replicatedCmd) error {
-	// NB: `origCmd` remains "Local". It's just not going to signal anyone
+// makeReproposal returns a new re-proposal of the given proposal, and a
+// function that must be called if this re-proposal is successfully proposed.
+//
+// We want to move a few items from origP to the new command, but only if we
+// managed to propose the new command. For example, if we move the latches over
+// too early but then fail to actually get the new proposal started, the old
+// proposal will not release the latches. This would result in a lost latch.
+func (r *Replica) makeReproposal(origP *ProposalData) (reproposal *ProposalData, success func()) {
+	// NB: original command remains "Local". It's just not going to signal anyone
 	// or release any latches.
 
-	origP := origCmd.proposal
 	seedP := origP.seedProposal
 	if seedP == nil {
 		seedP = origP
 	}
-
-	// We want to move a few items from origCmd to the new command, but only if we
-	// managed to propose the new command. For example, if we move the latches
-	// over too early but then fail to actually get the new proposal started, the
-	// old proposal will not release the latches. This would result in a lost
-	// latch.
-	var success bool
 
 	// Go through the original proposal field by field and decide what transfers
 	// to the new proposal (and how that affects the old proposal). The overall
@@ -356,10 +355,7 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
 		seedProposal: seedP,
 	}
 
-	defer func() {
-		if !success {
-			return
-		}
+	return newProposal, func() {
 		// If the original proposal had an explicit span, it's an async consensus
 		// proposal and the span would be finished momentarily (when we return to
 		// the caller) if we didn't unlink it here, but we want it to continue
@@ -390,7 +386,11 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
 		// We'd need to make sure the span is finished eventually.
 		origP.ctx = r.AnnotateCtx(context.TODO())
 		seedP.lastReproposal = newProposal
-	}()
+	}
+}
+
+func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *replicatedCmd) error {
+	newProposal, onSuccess := r.makeReproposal(origCmd.proposal)
 
 	// We need to track the request again in order to protect its timestamp until
 	// it gets reproposed.
@@ -424,7 +424,7 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
 	}
 	log.VEventf(ctx, 2, "reproposed command %x", newProposal.idKey)
 
-	success = true
+	onSuccess()
 	return nil
 }
 

--- a/pkg/kv/kvserver/replica_application_result.go
+++ b/pkg/kv/kvserver/replica_application_result.go
@@ -271,6 +271,10 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
 	// or release any latches.
 
 	origP := origCmd.proposal
+	seedP := origP.seedProposal
+	if seedP == nil {
+		seedP = origP
+	}
 
 	// We want to move a few items from origCmd to the new command, but only if we
 	// managed to propose the new command. For example, if we move the latches
@@ -365,6 +369,8 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
 		encodedCommand:          nil,
 		raftAdmissionMeta:       nil,
 		v2SeenDuringApplication: false,
+
+		seedProposal: seedP,
 	}
 	// If the original proposal had an explicit span, it's an async consensus
 	// proposal and the span would be finished momentarily (when we return to
@@ -377,6 +383,30 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
 		if success {
 			origP.sp = nil
 		}
+	}()
+
+	defer func() {
+		if !success {
+			return
+		}
+		// If the proposal is synchronous, the client is waiting on the seed
+		// proposal. By the time it has to act on the result, a bunch of reproposals
+		// can have happened, and some may still be running and using the
+		// context/tracing span (probably only the latest one, but we assume any,
+		// for defence-in-depth).
+		//
+		// Unbind the latest reproposal's context so that it no longer posts updates
+		// to the tracing span (it won't apply anyway). Link to the new latest
+		// reproposal, so that the client can clear its context at post-processing.
+		// This is effectively a "move" of the context to the reproposal.
+		//
+		// TODO(pavelkalinnikov): there should be a better way, after ProposalData
+		// lifecycle is reconsidered.
+		//
+		// TODO(radu): Should this context be created via tracer.ForkSpan?
+		// We'd need to make sure the span is finished eventually.
+		origP.ctx = r.AnnotateCtx(context.TODO())
+		seedP.lastReproposal = newProposal
 	}()
 
 	// We need to track the request again in order to protect its timestamp until

--- a/pkg/kv/kvserver/replica_application_result.go
+++ b/pkg/kv/kvserver/replica_application_result.go
@@ -320,11 +320,15 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
 		// for (some reincarnation of) the command to eventually apply, its trace
 		// will reflect the reproposal as well.
 		ctx:             origP.ctx,
-		sp:              origP.sp, // NB: special handling below
 		idKey:           raftlog.MakeCmdIDKey(),
 		proposedAtTicks: 0, // set in registerProposalLocked
 		createdAtTicks:  0, // set in registerProposalLocked
 		command:         &newCommand,
+
+		// Next comes the block of fields that are "moved" to the new proposal. See
+		// the deferred function call below which, correspondingly, clears these
+		// fields in the original proposal.
+		sp: origP.sp,
 		// NB: quotaAlloc is always nil here, because we already released the quota
 		// unconditionally in retrieveLocalProposals. So the below is a no-op.
 		//
@@ -333,8 +337,10 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
 		// releasing it here.
 		quotaAlloc: origP.quotaAlloc,
 		ec:         origP.ec,
-		applied:    false,
 		doneCh:     origP.doneCh,
+
+		applied: false,
+
 		// Local is copied over. It won't be used on the old proposal (since that
 		// proposal got rejected), but since it's still "local" we don't want to put
 		// it into  an undefined state by removing its response. The same goes for

--- a/pkg/kv/kvserver/replica_application_result.go
+++ b/pkg/kv/kvserver/replica_application_result.go
@@ -288,20 +288,6 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
 	// goal is that the old proposal remains a local proposal (switching it to
 	// non-local now invites logic bugs) but not bound to the caller.
 
-	// NB: quotaAlloc is always nil here, because we already
-	// released the quota unconditionally in retrieveLocalProposals.
-	// So the below is a no-op.
-	//
-	// TODO(tbg): if we shifted the release of proposal quota to *after*
-	// successful application, we could move the quota over
-	// prematurely releasing it here.
-	newQuotaAlloc := origP.quotaAlloc
-	defer func() {
-		if success {
-			origP.quotaAlloc = nil
-		}
-	}()
-
 	newCommand := kvserverpb.RaftCommand{
 		ProposerLeaseSequence: origP.command.ProposerLeaseSequence,
 		ReplicatedEvalResult:  origP.command.ReplicatedEvalResult,
@@ -316,25 +302,10 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
 		AdmissionOriginNode: 0,   // assigned on flush
 	}
 
-	// Now we construct the remainder of the ProposalData. First, the pieces
-	// that actively "move over", i.e. those that have to do with the latches
-	// held and the caller waiting to be signaled.
-
-	// `ec` (latches, etc) transfers to the new proposal.
-	newEC := origP.ec
-	defer func() {
-		if success {
-			origP.ec = makeEmptyEndCmds()
-		}
-	}()
-
-	// Ditto doneCh (signal to proposer).
-	newDoneCh := origP.doneCh
-	defer func() {
-		if success {
-			origP.doneCh = nil
-		}
-	}()
+	// Now we construct the remainder of the ProposalData. The pieces that
+	// actively "move over", are removed from the original proposal in the
+	// deferred func below. For example, those fields that have to do with the
+	// latches held and the caller waiting to be signaled.
 
 	// TODO(tbg): work on the lifecycle of ProposalData. This struct (and the
 	// surrounding replicatedCmd) are populated in an overly ad-hoc manner.
@@ -354,10 +325,16 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
 		proposedAtTicks: 0, // set in registerProposalLocked
 		createdAtTicks:  0, // set in registerProposalLocked
 		command:         &newCommand,
-		quotaAlloc:      newQuotaAlloc,
-		ec:              newEC,
-		applied:         false,
-		doneCh:          newDoneCh,
+		// NB: quotaAlloc is always nil here, because we already released the quota
+		// unconditionally in retrieveLocalProposals. So the below is a no-op.
+		//
+		// TODO(tbg): if we shifted the release of proposal quota to *after*
+		// successful application, we could move the quota over prematurely
+		// releasing it here.
+		quotaAlloc: origP.quotaAlloc,
+		ec:         origP.ec,
+		applied:    false,
+		doneCh:     origP.doneCh,
 		// Local is copied over. It won't be used on the old proposal (since that
 		// proposal got rejected), but since it's still "local" we don't want to put
 		// it into  an undefined state by removing its response. The same goes for
@@ -372,23 +349,23 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
 
 		seedProposal: seedP,
 	}
-	// If the original proposal had an explicit span, it's an async consensus
-	// proposal and the span would be finished momentarily (when we return to
-	// the caller) if we didn't unlink it here, but we want it to continue
-	// tracking newProposal. We leave it in `origP.ctx` though, since that
-	// context will become unused once the application of this (soft-failed)
-	// proposal concludes, i.e. soon after this method returns, in case there
-	// is anything left to log into it.
-	defer func() {
-		if success {
-			origP.sp = nil
-		}
-	}()
 
 	defer func() {
 		if !success {
 			return
 		}
+		// If the original proposal had an explicit span, it's an async consensus
+		// proposal and the span would be finished momentarily (when we return to
+		// the caller) if we didn't unlink it here, but we want it to continue
+		// tracking newProposal. We leave it in `origP.ctx` though, since that
+		// context will become unused once the application of this (soft-failed)
+		// proposal concludes, i.e. soon after this method returns, in case there is
+		// anything left to log into it.
+		origP.sp = nil
+		origP.quotaAlloc = nil
+		origP.ec = makeEmptyEndCmds()
+		origP.doneCh = nil
+
 		// If the proposal is synchronous, the client is waiting on the seed
 		// proposal. By the time it has to act on the result, a bunch of reproposals
 		// can have happened, and some may still be running and using the

--- a/pkg/kv/kvserver/replica_application_result_test.go
+++ b/pkg/kv/kvserver/replica_application_result_test.go
@@ -28,10 +28,7 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-func TestProposalDataAndRaftCommandAreConsideredWhenAddingFields(t *testing.T) {
-	defer leaktest.AfterTest(t)()
-	defer log.Scope(t).Close(t)
-
+func makeProposalData() *ProposalData {
 	raftCommand := &kvserverpb.RaftCommand{
 		ProposerLeaseSequence: 1,
 		MaxLeaseIndex:         1,
@@ -45,8 +42,8 @@ func TestProposalDataAndRaftCommandAreConsideredWhenAddingFields(t *testing.T) {
 		AdmissionOriginNode:   1,
 	}
 
-	prop := &ProposalData{
-		ctx:                     context.Background(),
+	return &ProposalData{
+		ctx:                     context.WithValue(context.Background(), struct{}{}, "nonempty-ctx"),
 		sp:                      &tracing.Span{},
 		idKey:                   "deadbeef",
 		proposedAtTicks:         1,
@@ -66,7 +63,13 @@ func TestProposalDataAndRaftCommandAreConsideredWhenAddingFields(t *testing.T) {
 		seedProposal:            nil,
 		lastReproposal:          nil,
 	}
+}
 
+func TestProposalDataAndRaftCommandAreConsideredWhenAddingFields(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	prop := makeProposalData()
 	// If you are adding a field to ProposalData or RaftCommand, please consider the
 	// desired semantics of that field in `tryReproposeWithNewLeaseIndex{,v2}`. Once
 	// this has been done, adjust the expected number of fields below, and populate
@@ -75,6 +78,49 @@ func TestProposalDataAndRaftCommandAreConsideredWhenAddingFields(t *testing.T) {
 	// NB: we can't use zerofields for two reasons: First, we have unexported fields
 	// here, and second, we don't want to check for recursively populated structs (but
 	// only for the top level fields).
-	require.Equal(t, 10, reflect.TypeOf(*raftCommand).NumField())
+	require.Equal(t, 10, reflect.TypeOf(*prop.command).NumField())
 	require.Equal(t, 19, reflect.TypeOf(*prop).NumField())
+}
+
+func TestReplicaMakeReproposalChaininig(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	var r Replica
+	proposals := make([]*ProposalData, 1, 4)
+	proposals[0] = makeProposalData()
+	sharedCtx := proposals[0].ctx
+
+	verify := func() {
+		seed := proposals[0]
+		require.Nil(t, seed.seedProposal)
+		// The seed proposal must know the latest reproposal.
+		if len(proposals) > 1 {
+			require.Equal(t, proposals[len(proposals)-1], seed.lastReproposal)
+		} else {
+			require.Nil(t, seed.lastReproposal)
+		}
+		// All reproposals must point at the seed proposal.
+		for _, reproposal := range proposals[1:] {
+			require.Equal(t, seed, reproposal.seedProposal)
+			require.Nil(t, reproposal.lastReproposal)
+		}
+		// Only the latest reproposal must use the seed context.
+		for _, prop := range proposals[:len(proposals)-1] {
+			require.NotEqual(t, sharedCtx, prop.ctx)
+		}
+		require.Equal(t, sharedCtx, proposals[len(proposals)-1].ctx)
+	}
+
+	verify()
+	for i := 1; i < cap(proposals); i++ {
+		reproposal, onSuccess := r.makeReproposal(proposals[i-1])
+		proposals = append(proposals, reproposal)
+		onSuccess()
+		verify()
+	}
+
+	reproposal, onSuccess := r.makeReproposal(proposals[len(proposals)-1])
+	_, _ = reproposal, onSuccess // No onSuccess call, assume the proposal failed.
+	verify()
 }

--- a/pkg/kv/kvserver/replica_application_result_test.go
+++ b/pkg/kv/kvserver/replica_application_result_test.go
@@ -63,6 +63,8 @@ func TestProposalDataAndRaftCommandAreConsideredWhenAddingFields(t *testing.T) {
 		tok:                     TrackedRequestToken{done: true},
 		raftAdmissionMeta:       &kvflowcontrolpb.RaftAdmissionMeta{},
 		v2SeenDuringApplication: true,
+		seedProposal:            nil,
+		lastReproposal:          nil,
 	}
 
 	// If you are adding a field to ProposalData or RaftCommand, please consider the
@@ -74,5 +76,5 @@ func TestProposalDataAndRaftCommandAreConsideredWhenAddingFields(t *testing.T) {
 	// here, and second, we don't want to check for recursively populated structs (but
 	// only for the top level fields).
 	require.Equal(t, 10, reflect.TypeOf(*raftCommand).NumField())
-	require.Equal(t, 17, reflect.TypeOf(*prop).NumField())
+	require.Equal(t, 19, reflect.TypeOf(*prop).NumField())
 }

--- a/pkg/kv/kvserver/replica_proposal.go
+++ b/pkg/kv/kvserver/replica_proposal.go
@@ -198,6 +198,23 @@ type ProposalData struct {
 	// application is). This flag makes sure that the proposal buffer won't
 	// accidentally reinsert a finished proposal into the map.
 	v2SeenDuringApplication bool
+
+	// seedProposal points at the seed proposal in the chain of (re-)proposals, if
+	// this ProposalData is a reproposal. The field is nil for the seed proposal.
+	seedProposal *ProposalData
+	// lastReproposal is the last proposal that superseded the seed proposal.
+	// Superseding proposals form a chain starting from the seed proposal. This
+	// field is set only on the seed proposal, and is updated every time a new
+	// reproposal is cast.
+	//
+	// See Replica.mu.proposals comment for more details.
+	//
+	// TODO(pavelkalinnikov): We are referencing only the last reproposal in the
+	// chain, so that the intermediate ones can be GCed in case the chain is very
+	// long. This chain reasoning is subtle, we should decompose ProposalData in
+	// such a way that the "common" parts of the (re-)proposals are shared and
+	// chaining isn't necessary.
+	lastReproposal *ProposalData
 }
 
 // useReplicationAdmissionControl indicates whether this raft command should

--- a/pkg/kv/kvserver/replica_raft.go
+++ b/pkg/kv/kvserver/replica_raft.go
@@ -325,9 +325,20 @@ func (r *Replica) evalAndPropose(
 		defer r.raftMu.Unlock()
 		r.mu.Lock()
 		defer r.mu.Unlock()
+		// When the caller abandons the request, it Finishes its trace. By that
+		// time, multiple reproposals can have occurred, and still running and
+		// attempting to post tracing updates through the context. This can cause a
+		// "use after Finish" race in the span. All the (re-)proposal contexts have
+		// been unbound except for the latest one. Unbind it to eliminate the race.
+		//
+		// See https://github.com/cockroachdb/cockroach/issues/107521
+		last := proposal
+		if p := proposal.lastReproposal; p != nil {
+			last = p
+		}
 		// TODO(radu): Should this context be created via tracer.ForkSpan?
 		// We'd need to make sure the span is finished eventually.
-		proposal.ctx = r.AnnotateCtx(context.TODO())
+		last.ctx = r.AnnotateCtx(context.TODO())
 	}
 	return proposalCh, abandon, idKey, writeBytes, nil
 }

